Index: wireless_tools.29/iwpriv.c
===================================================================
--- wireless_tools.29.orig/iwpriv.c
+++ wireless_tools.29/iwpriv.c
@@ -259,7 +259,7 @@ set_private_cmd(int		skfd,		/* Socket */
 		int		priv_num)	/* Number of descriptions */
 {
   struct iwreq	wrq;
-  u_char	buffer[4096];	/* Only that big in v25 and later */
+  u_char	buffer[20 * 1024];	/* Only that big in v25 and later */
   int		i = 0;		/* Start with first command arg */
   int		k;		/* Index in private description table */
   int		temp;
